src/System.Management.Automation/engine/parser/PSType.cs

```diff
@@ -1091,14 +1091,50 @@ internal static List<DefineEnumHelper> Sort(List<DefineEnumHelper> defineEnumHel
 
             internal void DefineEnum()
             {
-                var underlyingType = _enumDefinitionAst.BaseTypes.Count() == 0 ? typeof(int) : _enumDefinitionAst.BaseTypes[0].TypeName.GetReflectionType();
+                var typeConstraintAst = _enumDefinitionAst.BaseTypes.FirstOrDefault();
+                var underlyingType = typeConstraintAst == null ? typeof(int) : typeConstraintAst.TypeName.GetReflectionType();
 
                 var definedEnumerators = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
                 var enumBuilder = _moduleBuilder.DefineEnum(_typeName, Reflection.TypeAttributes.Public, underlyingType);
                 DefineCustomAttributes(enumBuilder, _enumDefinitionAst.Attributes, _parser, AttributeTargets.Enum);
 
                 dynamic value = 0;
-                dynamic maxValue = underlyingType.GetField("MaxValue", BindingFlags.Public | BindingFlags.Static | BindingFlags.GetField).GetValue(null);
+                dynamic maxValue = 0;
+                switch (Type.GetTypeCode(underlyingType))
+                {
+                    case TypeCode.Byte:
+                        maxValue = byte.MaxValue;
+                        break;
+                    case TypeCode.Int16:
+                        maxValue = short.MaxValue;
+                        break;
+                    case TypeCode.Int32:
+                        maxValue = int.MaxValue;
+                        break;
+                    case TypeCode.Int64:
+                        maxValue = long.MaxValue;
+                        break;
+                    case TypeCode.SByte:
+                        maxValue = sbyte.MaxValue;
+                        break;
+                    case TypeCode.UInt16:
+                        maxValue = UInt16.MaxValue;
+                        break;
+                    case TypeCode.UInt32:
+                        maxValue = UInt32.MaxValue;
+                        break;
+                    case TypeCode.UInt64:
+                        maxValue = ulong.MaxValue;
+                        break;
+                    default:
+                        _parser.ReportError(
```

Comment:
The same request - could we move the type check to parser?

Reply:
I honestly don't think that's a good idea. Injecting dynamic properties (we don't know the underlying type at compile time) into the AST - the AST being publicly exposed - sounds like asking for trouble. And we're emitting parser errors related to type constraints in the TypeDefiner for classes, it's not like this behavior is inconsistent with the existing implementation

Reply:
Enum looks like const. So we can make all checks at parse time.

Reply:
I know that we _can_, I'm saying we _shouldn't_ - if I apply [@lzybkr's line of reasoning here](https://github.com/IISResetMe/PowerShell/commit/bfd4be766246bdf33a0ffbb027ba66723fe06b49#r31450712), we'll at the very least have to keep the check in place here even if we also check at parse-time - and I'm not sure how useful that is

Reply:
Thanks for clarify! I see the point and agree that the design is good.